core/client/src/error.rs

```diff
@@ -16,159 +16,116 @@
+	UnknownBlock(String),
+	/// Applying extrinsic error.
+	#[display(fmt = "Extrinsic error: {:?}", _0)]
+	ApplyExtrinsicFailed(ApplyError),
+	/// Execution error.
+	#[display(fmt = "Execution: {}", _0)]
+	Execution(Box<state_machine::Error>),
+	/// Blockchain error.
+	#[display(fmt = "Blockchain: {}", _0)]
+	Blockchain(Box<std::error::Error + Send>),
+	/// Invalid authorities set received from the runtime.
+	#[display(fmt = "Current state of blockchain has invalid authorities set")]
+	InvalidAuthoritiesSet,
+	/// Could not get runtime version.
+	#[display(fmt = "On-chain runtime does not specify version")]
+	VersionInvalid,
+	/// Genesis config is invalid.
+	#[display(fmt = "Genesis config provided is invalid")]
+	GenesisInvalid,
+	/// Bad justification for header.
+	#[display(fmt = "bad justification for header: {}", _0)]
+	BadJustification(String),
+	/// Not available on light client.
+	#[display(fmt = "This method is not currently available when running in light client mode")]
+	NotAvailableOnLightClient,
+	/// Invalid remote CHT-based proof.
+	#[display(fmt = "Remote node has responded with invalid header proof")]
+	InvalidCHTProof,
+	/// Remote fetch has been cancelled.
+	#[display(fmt = "Remote data fetch has been cancelled")]
+	RemoteFetchCancelled,
+	/// Remote fetch has been failed.
+	#[display(fmt = "Remote data fetch has been failed")]
+	RemoteFetchFailed,
+	/// Error decoding call result.
+	#[display(fmt = "Error decoding call result of {}", _0)]
+	CallResultDecode(&'static str),
+	/// Error converting a parameter between runtime and node.
+	#[display(fmt = "Error converting `{}` between runtime and node", _0)]
+	RuntimeParamConversion(&'static str),
+	/// Changes tries are not supported.
+	#[display(fmt = "Changes tries are not supported by the runtime")]
+	ChangesTriesNotSupported,
+	/// Key changes query has failed.
+	#[display(fmt = "Failed to check changes proof: {}", _0)]
+	ChangesTrieAccessFailed(String),
+	/// Last finalized block not parent of current.
+	#[display(fmt = "Did not finalize blocks in sequential order.")]
+	NonSequentialFinalization(String),
+	/// Safety violation: new best block not descendent of last finalized.
+	#[display(fmt = "Potential long-range attack: block not in finalized chain.")]
+	NotInFinalizedChain,
+	/// Hash that is required for building CHT is missing.
+	#[display(fmt = "Failed to get hash of block#{} for building CHT#{}", _0, _1)]
+	MissingHashRequiredForCHT(u64, u64),
+	/// A convenience variant for String
+	#[display(fmt = "{}", _0)]
+	Msg(String),
+}
 
-		/// Hash that is required for building CHT is missing.
-		MissingHashRequiredForCHT(cht_num: u64, block_number: u64) {
-			description("missed hash required for building CHT"),
-			display("Failed to get hash of block#{} for building CHT#{}", block_number, cht_num),
+impl error::Error for Error {
+	fn source(&self) -> Option<&(error::Error + 'static)> {
+		match self {
+			Error::Consensus(e) => Some(e),
```

Comment:
Can we not add `Blockchain` and `Execution` here as well?

Reply:
@ascjones ?

Reply:
I've allowed `Blockchain` by changing it to the concrete type (it's recursive).

Not sure about `Execution` error, it's a custom trait and not `std::error::Error`. https://github.com/paritytech/substrate/blob/3a4901a78f627d689673597e5b8fca761d954ae3/core/state-machine/src/lib.rs#L61